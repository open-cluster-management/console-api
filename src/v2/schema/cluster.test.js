/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Cluster Resolver', () => {
  test('Correctly Resolves Cluster Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          clusters {
            clusterip
            metadata {
              creationTimestamp
              labels
              name
              namespace
              uid
            }
            availableVersions
            desiredVersion
            distributionVersion
            nodes
            status
            upgradeFailed
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Single Cluster Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          cluster(name: "hub-cluster", namespace: "hub-cluster") {
            clusterip
            metadata {
              creationTimestamp
              labels
              name
              namespace
              uid
            }
            availableVersions
            desiredVersion
            distributionVersion
            nodes
            status
            upgradeFailed
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves ClusterImageSet Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          clusterImageSets {
            name
            releaseImage
            channel
            visible
            platformAws
            platformGcp
            platformAzure
            platformBmc
            platformVmware
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves ClusterAddon Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          clusterAddons(name: "hub-cluster", namespace: "hub-cluster") {
            metadata {
              name
              namespace
            }
            addOnResource {
              name
              group
              resource
              description
            }
            status {
              message
              reason
              status
              type
            }
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});

describe('Cluster Mutation', () => {
  test('Correctly Resolves Update Cluster Labels', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          updateResourceLabels(resourceType:"HCMCompliance",apiVersion:"compliance.mcm.ibm.com/v1alpha1",kind:"Compliance",namespace:"my-remote-cluster-1", name:"compliance-xz", body:{cloud:"IBMs"}, selfLink:"/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/my-remote-cluster-1/compliances/compliance-xz", resourcePath:"/metadata/spec")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Create Cluster', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createCluster(cluster: [])
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Create Kubernetes Cluster Resource for Import with No Namespace', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            createCluster(
              cluster: [
                {
                  apiVersion: "cluster.open-cluster-management.io/v1",
                  kind: "ManagedCluster"
                },
                {
                  apiVersion: "agent.open-cluster-management.io/v1",
                  kind: "KlusterletAddonConfig",
                  metadata:
                  {
                    name: "foo",
                    namespace: "foo"
                  },
                  spec:
                  {
                    clusterName: "foo"
                  }
                }
              ]
            )
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.error)).toBeFalsy();
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Create Kubernetes Cluster Resource for Import with Import YAML Problem', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            createCluster(
              cluster: [
                {
                  apiVersion: "cluster.open-cluster-management.io/v1",
                  kind: "ManagedCluster"
                  metadata:
                  {
                    name: "no-yaml"
                  }
                },
                {
                  apiVersion: "agent.open-cluster-management.io/v1",
                  kind: "KlusterletAddonConfig",
                  metadata:
                  {
                    name: "no-yaml",
                    namespace: "no-yaml"
                  },
                  spec:
                  {
                    clusterName: "no-yaml"
                  }
                }
              ]
            )
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.error)).toBeFalsy();
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Detach Cluster', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          detachCluster(namespace:"default", cluster:"managed-cluster")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Detach Cluster with Failure', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          detachCluster(namespace:"kube-system", cluster:"hub-cluster")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Destroy Cluster', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          detachCluster(namespace:"default", cluster:"managed-cluster", destroy: true)
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Destroy Cluster with Failure', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          detachCluster(namespace:"kube-system", cluster:"new-cluster", destroy: true)
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});
