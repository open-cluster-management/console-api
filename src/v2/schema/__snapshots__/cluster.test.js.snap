// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cluster Mutation Correctly Resolves Create Cluster 1`] = `
Object {
  "data": Object {
    "createCluster": Object {
      "errors": Array [
        Object {
          "message": "Cannot find any endpoints",
        },
      ],
    },
  },
}
`;

exports[`Cluster Mutation Correctly Resolves Destroy Cluster 1`] = `
Object {
  "data": Object {
    "detachCluster": 204,
  },
}
`;

exports[`Cluster Mutation Correctly Resolves Destroy Cluster with Failure 1`] = `
Object {
  "data": Object {
    "detachCluster": Object {
      "code": "406",
      "kind": "Status",
      "status": "Not Acceptable",
    },
  },
}
`;

exports[`Cluster Mutation Correctly Resolves Detach Cluster 1`] = `
Object {
  "data": Object {
    "detachCluster": 204,
  },
}
`;

exports[`Cluster Mutation Correctly Resolves Detach Cluster with Failure 1`] = `
Object {
  "data": Object {
    "detachCluster": Object {
      "code": "401",
      "kind": "Status",
    },
  },
}
`;

exports[`Cluster Mutation Correctly Resolves Update Cluster Labels 1`] = `
Object {
  "data": Object {
    "updateResourceLabels": Object {
      "apiVersion": "compliance.mcm.ibm.com/v1alpha1",
      "kind": "Compliance",
      "metadata": Object {
        "annotations": Object {
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-01-02T19:42:12Z",
        "finalizers": Array [
          "finalizer.mcm.ibm.com",
        ],
        "generation": 1,
        "name": "compliance-all",
        "namespace": "mcm",
        "resourceVersion": "5533372",
        "selfLink": "/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/mcm/compliances/compliance-all",
        "uid": "7fc1f4a2-0ec6-11e9-8fd0-0ebe277f4f9c",
      },
      "spec": Object {
        "runtime-rules": Array [
          Object {
            "apiVersion": "policy.open-cluster-management.io/v1",
            "kind": "Policy",
            "metadata": Object {
              "creationTimestamp": null,
              "labels": Object {
                "hipaa": "true",
              },
              "name": "policy-all",
            },
            "spec": Object {
              "complianceType": "musthave",
              "namespaces": Object {
                "exclude": Array [
                  "kube*",
                ],
                "include": Array [
                  "default",
                ],
              },
              "object-templates": Array [
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "Pod",
                    "metadata": Object {
                      "name": "nginx",
                    },
                    "spec": Object {
                      "containers": Array [
                        Object {
                          "image": "nginx:1.7.9",
                          "name": "nginx",
                          "ports": Array [
                            Object {
                              "containerPort": 80,
                            },
                          ],
                        },
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "Namespace",
                    "metadata": Object {
                      "labels": Object {
                        "name": "production",
                      },
                      "name": "production",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "rbac.authorization.k8s.io/v1",
                    "kind": "RoleBinding",
                    "metadata": Object {
                      "name": "operate-pods-rolebinding",
                      "namespace": "default",
                    },
                    "roleRef": Object {
                      "apiGroup": "rbac.authorization.k8s.io",
                      "kind": "Role",
                      "name": "operator",
                    },
                    "subjects": Array [
                      Object {
                        "apiGroup": "rbac.authorization.k8s.io",
                        "kind": "User",
                        "name": "jane",
                      },
                    ],
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "policy/v1beta1",
                    "kind": "PodSecurityPolicy",
                    "metadata": Object {
                      "annotations": Object {
                        "seccomp.security.alpha.kubernetes.io/allowedProfileNames": "*",
                      },
                      "name": "privileged-mcm",
                    },
                    "spec": Object {
                      "allowPrivilegeEscalation": true,
                      "allowedCapabilities": Array [
                        "*",
                      ],
                      "fsGroup": Object {
                        "rule": "RunAsAny",
                      },
                      "hostIPC": true,
                      "hostNetwork": true,
                      "hostPID": true,
                      "hostPorts": Array [
                        Object {
                          "max": 65535,
                          "min": 0,
                        },
                      ],
                      "privileged": true,
                      "runAsUser": Object {
                        "rule": "RunAsAny",
                      },
                      "seLinux": Object {
                        "rule": "RunAsAny",
                      },
                      "supplementalGroups": Object {
                        "rule": "RunAsAny",
                      },
                      "volumes": Array [
                        "*",
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "networking.k8s.io/v1",
                    "kind": "NetworkPolicy",
                    "metadata": Object {
                      "name": "allow-all-mcm",
                    },
                    "spec": Object {
                      "ingress": Array [
                        Object {},
                      ],
                      "podSelector": Object {},
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "LimitRange",
                    "metadata": Object {
                      "name": "mem-limit-range",
                    },
                    "spec": Object {
                      "limits": Array [
                        Object {
                          "default": Object {
                            "memory": "512Mi",
                          },
                          "defaultRequest": Object {
                            "memory": "256Mi",
                          },
                          "type": "Container",
                        },
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
              "remediationAction": "enforce",
              "role-templates": Array [
                Object {
                  "apiVersion": "roletemplate.mcm.ibm.com/v1alpha1",
                  "complianceType": "musthave",
                  "metadata": Object {
                    "creationTimestamp": null,
                    "name": "operator-role",
                  },
                  "rules": Array [
                    Object {
                      "complianceType": "musthave",
                      "policyRule": Object {
                        "apiGroups": Array [
                          "extensions",
                          "apps",
                        ],
                        "resources": Array [
                          "deployments",
                        ],
                        "verbs": Array [
                          "get",
                          "list",
                          "watch",
                          "create",
                          "delete",
                          "patch",
                        ],
                      },
                    },
                  ],
                  "selector": Object {
                    "matchLabels": Object {
                      "hipaa": "true",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
            },
            "status": Object {},
          },
        ],
      },
      "status": Object {
        "placementBindings": Array [
          "binding-xz",
        ],
        "placementPolicies": Array [
          "placement-xz",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster1",
            "compliant": "Compliant",
          },
          "cluster2": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster2",
            "compliant": "Compliant",
          },
          "cluster3": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster3",
            "compliant": "Compliant",
          },
        },
      },
    },
  },
}
`;

exports[`Cluster Mutation Create Kubernetes Cluster Resource for Import with Import YAML Problem 1`] = `
Object {
  "data": Object {
    "createCluster": Object {
      "created": Array [
        Object {
          "kind": "Cluster",
          "name": "a-cluster",
        },
        Object {
          "kind": "KlusterletAddonConfig",
        },
      ],
      "errors": Array [
        Object {
          "message": "secrets \\"no-yaml-import\\" not found",
        },
      ],
      "importSecret": Object {
        "apiVersion": "v1",
        "code": 404,
        "details": Object {
          "kind": "secrets",
          "name": "no-yaml-import",
        },
        "kind": "Status",
        "message": "secrets \\"no-yaml-import\\" not found",
        "metadata": Object {},
        "reason": "NotFound",
        "status": "Failure",
      },
      "updated": Array [],
    },
  },
}
`;

exports[`Cluster Mutation Create Kubernetes Cluster Resource for Import with No Namespace 1`] = `
Object {
  "data": Object {
    "createCluster": Object {
      "errors": Array [
        Object {
          "message": "No namespace specified",
        },
      ],
    },
  },
}
`;

exports[`Cluster Resolver Correctly Resolves Cluster Query 1`] = `
Object {
  "data": Object {
    "clusters": Array [
      Object {
        "availableVersions": null,
        "clusterip": "9.42.23.230",
        "desiredVersion": null,
        "distributionVersion": null,
        "metadata": Object {
          "creationTimestamp": "2018-08-13T16:56:41Z",
          "labels": Object {
            "cloud": "IBM",
            "datacenter": "toronto",
            "environment": "Dev",
            "name": "managed-cluster",
            "owner": "marketing",
            "region": "US",
            "vendor": "ICP",
          },
          "name": "managed-cluster",
          "namespace": "managed-cluster",
          "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
        },
        "nodes": 2,
        "status": "pendingimport",
        "upgradeFailed": null,
      },
      Object {
        "availableVersions": Array [
          "4.2.8",
          "4.2.7",
          "4.2.2",
          "4.2.9",
        ],
        "clusterip": "9.42.23.230",
        "desiredVersion": "4.2.2",
        "distributionVersion": "4.2.0",
        "metadata": Object {
          "creationTimestamp": "2018-08-13T16:56:41Z",
          "labels": Object {
            "cloud": "IBM",
            "datacenter": "raleigh",
            "environment": "Dev",
            "name": "hub-cluster",
            "owner": "development",
            "region": "US",
            "vendor": "ICP",
          },
          "name": "hub-cluster",
          "namespace": "hub-cluster",
          "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
        },
        "nodes": 3,
        "status": "ok",
        "upgradeFailed": false,
      },
      Object {
        "availableVersions": null,
        "clusterip": "9.42.23.230",
        "desiredVersion": null,
        "distributionVersion": null,
        "metadata": Object {
          "creationTimestamp": "2018-08-13T16:56:41Z",
          "labels": Object {
            "cloud": "IBM",
            "datacenter": "raleigh",
            "environment": "Dev",
            "name": "new-cluster",
            "owner": "development",
            "region": "US",
            "vendor": "ICP",
          },
          "name": "new-cluster",
          "namespace": "new-cluster",
          "uid": "d9f3a6aa-9f19-11f8-855e-f2b908610543",
        },
        "nodes": 4,
        "status": "offline",
        "upgradeFailed": null,
      },
    ],
  },
}
`;

exports[`Cluster Resolver Correctly Resolves ClusterImageSet Query 1`] = `
Object {
  "data": Object {
    "clusterImageSets": Array [
      Object {
        "channel": "fast",
        "name": "img4.3.0-x86-64",
        "platformAws": "true",
        "platformAzure": "true",
        "platformBmc": null,
        "platformGcp": "true",
        "platformVmware": null,
        "releaseImage": "quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64",
        "visible": "true",
      },
      Object {
        "channel": null,
        "name": "img4.4.6-x86-64",
        "platformAws": null,
        "platformAzure": null,
        "platformBmc": null,
        "platformGcp": null,
        "platformVmware": null,
        "releaseImage": "quay.io/openshift-release-dev/ocp-release:4.4.6-x86_64",
        "visible": null,
      },
      Object {
        "channel": "stable",
        "name": "img4.4.8-x86-64",
        "platformAws": "true",
        "platformAzure": "true",
        "platformBmc": "true",
        "platformGcp": "true",
        "platformVmware": "true",
        "releaseImage": "quay.io/openshift-release-dev/ocp-release:4.4.8-x86_64",
        "visible": "false",
      },
    ],
  },
}
`;

exports[`Cluster Resolver Correctly Resolves Single Cluster Query 1`] = `
Object {
  "data": Object {
    "cluster": Array [
      Object {
        "availableVersions": Array [
          "4.2.8",
          "4.2.7",
          "4.2.2",
          "4.2.9",
        ],
        "clusterip": "9.42.23.230",
        "desiredVersion": "4.2.2",
        "distributionVersion": "4.2.0",
        "metadata": Object {
          "creationTimestamp": "2018-08-13T16:56:41Z",
          "labels": Object {
            "cloud": "IBM",
            "datacenter": "raleigh",
            "environment": "Dev",
            "name": "hub-cluster",
            "owner": "development",
            "region": "US",
            "vendor": "ICP",
          },
          "name": "hub-cluster",
          "namespace": "hub-cluster",
          "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
        },
        "nodes": 3,
        "status": "ok",
        "upgradeFailed": false,
      },
    ],
  },
}
`;
